compiler/p/codegen/OMRCodeGenerator.cpp

```diff
@@ -1713,6 +1705,22 @@ void OMR::Power::CodeGenerator::generateBinaryEncodingPrologue(
    self()->getLinkage()->createPrologue(data->cursorInstruction);
    }
 
+static void expandFarConditionalBranches(TR::CodeGenerator *cg)
+   {
+   for (TR::Instruction *instr = cg->getFirstInstruction(); instr; instr = instr->getNext())
+      {
+      TR::PPCConditionalBranchInstruction *branch = instr->getPPCConditionalBranchInstruction();
+
+      if (branch && branch->getLabelSymbol())
+         {
+         int32_t distance = branch->getLabelSymbol()->getEstimatedCodeLocation() - branch->getEstimatedBinaryLocation();
+
+         if (distance < (int32_t)TR::getMinSigned<TR::Int16>() || distance > (int32_t)TR::getMaxSigned<TR::Int16>())
```

Comment:
this implementation seemed increasing the fragility of code correctness. it appeared you will depend on the cutting-edge precision of the instruction location estimation, otherwise you might miss potential far conditional branches. But, by estimation, you are not at cutting-edge precision. That was why the old implementation was conservative on forward conditional branches (while backward branches have been precise).

Reply:
Another way of looking at this, by estimating 2x length, you are increasing the likelihood of far conditional branches unnecessarily, especially for methods with length below 32KB.

Reply:
I'm not sure I follow your reasoning regarding the fragility of code correctness. As I understand it, unless there's something I'm missing here, estimating conditional branches at double length should mean that all estimated instruction lengths will be an upper bound, meaning that the actual (absolute) distance between the branch and its target should never be higher than the estimate calculated here.

While it's true that doing this will cause branches to be estimated more conservatively, IMO (and as pointed out by @ymanton in his review) this reduction in complexity should be worth it, as the old code was very difficult to understand and poorly documented. Additionally, simplifying this code allows me to add asserts that ensure that no `TR::Instruction` ever encodes as being longer than its original estimate (this will be done in tandem with the rest of the binary encoding refactor), which should dramatically reduce the likelihood of any bugs in related code by causing test failures more readily and ensure that any bugs in binary length estimation should manifest in a consistent, easy to debug manner.

All that being said, this is something I'd like to revisit optimizing in a future PR in a more clear and documented manner that still allows the previously mentioned asserts to work correctly. If you're okay with it, then I'll go ahead and open an issue to that effect once this PR is merged.

Reply:
I am fine with this stance ... sign off.